File: src/lib/types/product.ts
export interface ProductComponent {
  component_id: string;
  description: string;
  quantity_per: number;
  material_cost: number;
  labour_cost: number;
  supplier: string;
  date_created: string;
  unallocated: number;
}

export interface SalesOrder {
  sales_order_id: string;
  location: string;
  customer: string;
  order_date: string | null;
  completion_date: string | null;
  order_quantity: number;
  build_cost: number;
  sale_price: number;
  percent_complete: number;
}

export interface ProductData {
  product_id: string;
  description: string;
  components: ProductComponent[];
  sales_orders: SalesOrder[];
}

File: src-tauri/src/schema/product.rs
use super::deserialize_date;

#[derive(Debug, serde::Serialize, serde::Deserialize)]
pub struct ProductData {
    product_id: String,
    description: String,
    #[serde(default)]
    components: Vec<ProductComponent>,
    #[serde(default)]
    sales_orders: Vec<SalesOrder>,
}

#[derive(Debug, serde::Serialize, serde::Deserialize)]
struct ProductComponent {
    component_id: String,
    description: String,
    quantity_per: f64,
    material_cost: f64,
    labour_cost: f64,
    supplier: String,
    date_created: String,
    unallocated: f64,
}

#[derive(Debug, serde::Serialize, serde::Deserialize)]
struct SalesOrder {
    sales_order_id: String,
    location: String,
    customer: String,
    #[serde(default, deserialize_with = "deserialize_date")]
    order_date: Option<String>,
    #[serde(default, deserialize_with = "deserialize_date")]
    completion_date: Option<String>,
    order_quantity: f64,
    build_cost: f64,
    sale_price: f64,
    percent_complete: f64,
}
